rfcs/2025-03-16-smp-queues.md

```diff
@@ -0,0 +1,163 @@
+    -- connReq including the full link allows connection redundancy.
+    -- Technically, it would allow clients to point to queue on another server.
+    -- But clients should reject addresses when the primary queue is on a different server.
+    -- Clients could also warn when any of the queues is on unknown operator.
+    userData :: ByteString -- the max size needs to be estimated, but it is likely to be ~ 14kb
+  }
+
+-- | Updated queue IDs and keys, returned in IDS response
+data QueueIdsKeys = QIK
+  { rcvId :: RecipientId,
+    linkId :: Maybe LinkId, -- same as in ShortLink
+    sndId :: SenderId, -- same as in ShortLink or random
+    rcvPublicDhKey :: RcvPublicDhKey,
+    sndSecure :: SenderCanSecure, -- possibly, can be removed? or implied?
+    serverNtfCreds :: Maybe ServerNtfCreds -- currently returned in NID response
+  }
+
+data ServerNtfCreds = ServerNtfCreds NotifierId RcvNtfPublicDhKey -- same NotifierId as in NtfRequest
+```
+
+There was a consideration that the full invitation link can be hosted on a different server, but this is not proposed, as while it can potentially improve transport anonymity, it reduces reliability and also may reduce transport privacy of accepting party.
+
+In addition to that we would add the command allowing to update and also to retrieve and, optionally, secure the queue sender ID and data in one request, to reduce rountrips:
+
+```haskell
+LSET :: EncLinkData -> Command Recipient
+
+-- with LinkId as entity ID, if queue does not allow sndSecure the command will fail, same as SKEY
+LGET :: Maybe SndPublicAuthKey -> Command Sender
+
+LINK :: QueueLink -> BrokerMsg
+```
+
+## Algorithm to prepare and to interpret queue link data.
+
+This follows the design proposed in [Short links](./2024-06-21-short-links.md) RFC.
+
+**Prepare queue link data**
+
+- the data blob owner generates X25519 key pair: `(k, pk)`.
+- private key `pk` will be included in the short link shared with the other party (only base64url encoded key bytes, not X509 encoding).
+- `HKDF(pk)` will be used to encrypt the link data with secret_box before storing it on the server.
+- the hash of public key `sha256(k)` will be used as ID by the owner to update the link data and by the accepting party to get link (`LSET` and `LGET` commands).
+
+**Retrieving queue link data**
+
+- the sender uses the public key `k` derived from the private key `pk` included in the link as entity ID to retrieve link data (the server will compute the ID used by the owner as `sha256(k)` and will be able to look it up). This provides the quality that the traffic of the parties has no shared IDs inside TLS. It also means that unlike message queue creation, the ID to retrieve the link data was never sent to the creator, and also is not known to the server in advance (the second part is only an observation, in itself it does not increase security, as server has access to an encrypted blob anyway).
```

Comment:
the first part also does not increase security, as way to compute id from key is known

Reply:
which part?